File: bubsub/src/lib/money.ts
import type { Subscription } from './types'
import { convert } from './currency'

const CYCLES_PER_MONTH: Record<Subscription['cycle'], number> = {
  weekly: 52 / 12,
  monthly: 1,
  quarterly: 1 / 3,
  yearly: 1 / 12,
}

/** Price of one subscription normalized to a monthly amount, in the display currency. */
export function monthlyCost(sub: Subscription, displayCurrency: string): number {
  const monthly = sub.price * CYCLES_PER_MONTH[sub.cycle]
  return convert(monthly, sub.currency, displayCurrency)
}

export function yearlyCost(sub: Subscription, displayCurrency: string): number {
  return monthlyCost(sub, displayCurrency) * 12
}

export function isCounted(sub: Subscription): boolean {
  return sub.status === 'active' || sub.status === 'trial'
}

export function totalMonthly(subs: Subscription[], displayCurrency: string): number {
  return subs.filter(isCounted).reduce((sum, s) => sum + monthlyCost(s, displayCurrency), 0)
}

export function totalYearly(subs: Subscription[], displayCurrency: string): number {
  return totalMonthly(subs, displayCurrency) * 12
}

/** Yearly savings from everything the user already cancelled. */
export function cancelledSavings(subs: Subscription[], displayCurrency: string): number {
  return subs
    .filter((s) => s.status === 'cancelled')
    .reduce((sum, s) => sum + yearlyCost(s, displayCurrency), 0)
}

File: app/lib/data/money.dart
import 'models.dart';

const usdRates = {
  'USD': 1.0,
  'EUR': 0.92,
  'GBP': 0.79,
  'EGP': 48.5,
  'PLN': 3.95,
  'SAR': 3.75,
  'AED': 3.67,
  'KWD': 0.31,
  'QAR': 3.64,
  'TRY': 34.5,
  'INR': 84.0,
  'CAD': 1.38,
  'AUD': 1.52,
  'JPY': 152.0,
};

double convert(double amount, String from, String to) {
  if (from == to) return amount;
  final fromRate = usdRates[from] ?? 1;
  final toRate = usdRates[to] ?? 1;
  return (amount / fromRate) * toRate;
}

double monthlyCost(Subscription s, String display) {
  final factor = switch (s.cycle) {
    BillingCycle.weekly => 52 / 12,
    BillingCycle.monthly => 1.0,
    BillingCycle.quarterly => 1 / 3,
    BillingCycle.yearly => 1 / 12,
  };
  return convert(s.price * factor, s.currency, display);
}

bool isCounted(Subscription s) => s.status == SubStatus.active || s.status == SubStatus.trial;

double totalMonthly(List<Subscription> list, String display) =>
    list.where(isCounted).fold(0.0, (a, s) => a + monthlyCost(s, display));

String formatMoney(double amount, String currency) {
  final symbols = {
    'USD': '\$',
    'EUR': '€',
    'GBP': '£',
    'EGP': 'E£',
    'PLN': 'zł',
    'SAR': 'SR',
    'AED': 'AED',
    'KWD': 'KD',
    'QAR': 'QR',
    'TRY': '₺',
    'INR': '₹',
    'CAD': 'C\$',
    'AUD': 'A\$',
    'JPY': '¥',
  };
  final sym = symbols[currency] ?? '$currency ';
  // Always show cents for money clarity (avoids "$15" looking rounded/broken).
  return '$sym${amount.toStringAsFixed(2)}';
}
